### DISCUSSION TITLE
How to properly mock an API in testing?

### DISCUSSION THREAD

**Discussion by MichaelHochriegl** [jasontaylordev/CleanArchitecture#393] [Q&A] [answered]
Hey all,

I'm really liking this template, but I'm right now struggling on how to best mock an external API my app is using for testing.
I have an interface that my ApiClient inherits, so that I can mock it, but I don't know how I would best use the mock.

In my current mindset I would need to define the mock-method used in almost every testcase differently (testing if appropriate data is returned from the API, testing if nothing is returned from the API ...), but how do I register this specific mock with the DI so that the mediator handler is going to use it?

I hope that my problem is understandable. If my current though-process on how to tackle this is wrong than I'm more than happy to learn.

**Comment by ronnieholm** [comment, marked as answer]
I've extended `TestBase.cs` and `Testing.cs`. Here's an example of mocking `IEmailSender`:

https://gist.github.com/ronnieholm/7c97b09097c44a3c33a7f36c0d053fff

**Reply by fakhrulhilal** [reply]
Beware when using checkpoint. Basically it truncate all tables except what we've configured. When we use huge database, it will take a long time to clean up database, and perhaps, you may lose data that you don't expect. I suggest to use dedicated database for testing only. Perhaps, the database will be built before running the whole test, and dropped when finished.  So there will be no issue for this.

**Comment by MichaelHochriegl** [comment]
Thanks for the fast answer, @ronnieholm ! I too thought about your solution to this problem, but wouldn't this recreate the entire DI container for each test? Is this something to even worry about?
It's "only" testing and not production, but if there would be any singletons the would exist multiple times.
